### ISSUE TITLE
Sending JSON data hangs when ghc was built with integer-simple.

### ISSUE THREAD

**Issue by kantp** [yesodweb/yesod#981]
Hi,

I have encountered some strange behaviour in an app I have put together using Yesod. I'm not perfectly sure whether the problem is actually in Yesod or somewhere deeper.

It seems that whenever I try to serve JSON data containing numeric types from a Yesod App that was compiled using ghc with integer-simple instead of integer-gmp, the App just hangs, using ~ 100% CPU.

I've put together a minimal test case (following loosely http://www.yesodweb.com/book/restful-content):

```
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes       #-}
{-# LANGUAGE TemplateHaskell   #-}
{-# LANGUAGE TypeFamilies      #-}

module Main where

import           Data.Aeson (encode)
import           Yesod

data WrappedInt = WrappedInt Int
instance ToJSON WrappedInt where
  toJSON (WrappedInt x) = object [ "WrappedInt" .= x]

val :: WrappedInt
val = WrappedInt 42

data App = App

mkYesod "App" [parseRoutes|
/ HomeR GET
|]

instance Yesod App

getHomeR :: Handler Value
getHomeR =
  liftIO $ print (encode val)
  >> returnJson val

main :: IO ()
main = warp 3000 App
```

If I run this (using ghc with integer-simple) and go to http://localhost:3000, it prints the encoded value to the terminal (from print), but the client waits forever for the reply. Since printing works, I don't think the issue is with the aeson library.

Switching to another integer datatype like Int64 or Integer does not resolve the problem. Using a String, or ghc compiled with integer-gmp, does.

Do you have any idea what could cause this? If you need more info, please let me know.

Best,
Philipp

**Comment by snoyberg** [comment]
That's an interesting bug. Can you try just rendering the json to a lazy byte string and see if that reproduces the problem?

**Comment by kantp** [comment]
Yes, interesting is the right word, I guess ;-)

Re. rendering as a lazy byte string: I think that's what I did with liftIO $ print (encode val). Or did you mean rendering to a lazy byte string and then sending that as a response?

**Comment by snoyberg** [comment]
I meant a test case that didn't use Yesod at all, just aeson, to really try and pinpoint where the issue is coming from.

**Comment by kantp** [comment]
Ah, ok. Yes, I tried that, and it worked.

**Comment by snoyberg** [comment]
I'm not sure what would be causing this. One guess is that there's some kind of a bug in the Builder implementation with integer-simple. I'm not going to be able to dig into this myself right now. The best path forward would be to play around with other response formats (such as other lazy `ByteString`s generated in different ways) and see if anything else reproduces this problem.

**Comment by kantp** [comment]
Thanks, I'll try that. If I find anything, I'll let you know.

**Comment by andrewthad** [comment]
The issue is not in yesod. It's in the `text` package. Here are some links that better explain this problem:

  - Issue with yesod: https://github.com/yesodweb/yesod/issues/967
  - Issue with aeson: https://github.com/bos/aeson/issues/246
  - Issue with text (the real culprit): https://github.com/bos/text/issues/114

It seems like it might be a while before this gets fixed. The issue against `text` was opened two weeks ago, and I don't think that very many people use `integer-simple`, so there aren't exactly a bunch of us affected by this.

As an workaround for now, we can take advantage of the fact that the bytestring implementation is not broken. I have a helper function that looks like this:

    encodeToText :: ToJSON a => a -> Text
    encodeToText = TE.decodeUtf8 . LByteString.toStrict . encode                                          

It uses a partial function, but it does the job for now. I really hope this gets fixed.

**Comment by kantp** [comment]
Thanks a lot for this info, Andrew!

**Comment by kantp** [comment]
With the recent fix in the text library (bos/text#117), this issue is resolved, so I'm closing this.

Thanks to everyone involved!
